helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java

```diff
@@ -0,0 +1,562 @@
+      removingIdealStates.keySet().removeAll(newIdealStateMap.keySet());
+      for (String resource : removingIdealStates.keySet()) {
+        _admin.dropResource(_clusterName, resource);
+      }
+      // update the new resource idealStates
+      for (IdealState is : newIdealStateMap.values()) {
+        _admin.setResourceIdealState(_clusterName, is.getResourceName(), is);
+      }
+      idealStateMap = newIdealStateMap;
+
+      // Override resource config
+      resources = _admin.getResourcesInCluster(_clusterName);
+      Map<String, ResourceConfig> resourceConfigMap = resources.stream().map(resource -> {
+        ResourceConfig config = configAccessor.getResourceConfig(_clusterName, resource);
+        if (config == null) {
+          config = new ResourceConfig(resource);
+        }
+        return config;
+      }).collect(Collectors.toMap(resourceConfig -> resourceConfig.getResourceName(),
+          resourceConfig -> resourceConfig));
+      Map<String, ResourceConfig> newResourceConfigMap =
+          overwrites.updateResourceConfigs(resourceConfigMap, newIdealStateMap);
+      // remove all the non-exist resources
+      Map<String, ResourceConfig> removingResourceConfigs = new HashMap<>(resourceConfigMap);
+      removingResourceConfigs.keySet().removeAll(newResourceConfigMap.keySet());
+      for (String resource : removingResourceConfigs.keySet()) {
+        HelixConfigScope resourceScope =
+            new HelixConfigScopeBuilder(HelixConfigScope.ConfigScopeProperty.RESOURCE)
+                .forCluster(_clusterName).forResource(resource).build();
+        configAccessor.remove(resourceScope, resource);
+      }
+      // update the new resource configs
+      for (ResourceConfig config : newResourceConfigMap.values()) {
+        configAccessor.setResourceConfig(_clusterName, config.getResourceName(), config);
+      }
+    }
+
+    ClusterConfig clusterConfig = configAccessor.getClusterConfig(_clusterName);
+    clusterConfig.setPersistBestPossibleAssignment(true);
+    configAccessor.setClusterConfig(_clusterName, clusterConfig);
+
+    for (IdealState is : idealStateMap.values()) {
+      _admin.rebalance(_clusterName, is.getResourceName(),
+          is.getReplicaCount(instanceConfigMap.size()));
+    }
+
+    for (InstanceConfig instanceConfig : instanceConfigMap.values()) {
+      startNewProcess(instanceConfig.getInstanceName(), stateModelDefRefs);
+      _nodeToZoneMap.put(instanceConfig.getInstanceName(), instanceConfig.getZoneId());
+    }
+  }
+
+  private void executeScript(List<Operation> operations) throws IOException {
+    for (Operation operation : operations) {
+      echo("Start operation: " + operation.getDescription());
+      // Reset the count before operation.
+      for (SimulateParticipantProcess process : _processMap.values()) {
+        process.getAndResetTransitions();
+      }
+      if (operation.execute(_admin, _zkClient, _clusterName)) {
+
+        ClusterStateVerifier.verifyByPolling(() -> {
```

Comment:
Just add some comments here that we want to see at least 1 transition happening...